File: src/core/health.ts
export interface HealthSample {
  latency_ms: number | null;
  packet_loss_pct: number | null;
  link_quality: number | null;
}

export function healthScore(sample: HealthSample): number {
  const latency =
    sample.latency_ms === null || sample.latency_ms === undefined
      ? 0.0
      : Math.max(0.0, 1.0 - sample.latency_ms / 500.0);

  const loss =
    sample.packet_loss_pct === null || sample.packet_loss_pct === undefined
      ? 0.0
      : Math.max(0.0, 1.0 - sample.packet_loss_pct / 100.0);

  const quality =
    sample.link_quality === null || sample.link_quality === undefined
      ? 0.0
      : Math.max(0.0, Math.min(1.0, sample.link_quality));

  return latency * 0.35 + loss * 0.35 + quality * 0.3;
}

File: src/wifi_framework/health.py
from dataclasses import dataclass


@dataclass
class HealthSample:
    latency_ms: float | None
    packet_loss_pct: float | None
    link_quality: float | None


def health_score(sample: HealthSample) -> float:
    latency = 0.0 if sample.latency_ms is None else max(0.0, 1.0 - sample.latency_ms / 500.0)
    loss = 0.0 if sample.packet_loss_pct is None else max(0.0, 1.0 - sample.packet_loss_pct / 100.0)
    quality = 0.0 if sample.link_quality is None else max(0.0, min(1.0, sample.link_quality))
    return latency * 0.35 + loss * 0.35 + quality * 0.30
